File: packages/vscode-extension/src/client/SessionManager.ts
import { WsClient } from './WsClient';
import { JiuwenMessage, SessionInfo, makeRequest } from './protocol';

type PendingRequest = {
  resolve: (payload: Record<string, unknown>) => void;
  reject: (err: Error) => void;
  timer: NodeJS.Timeout;
};

const REQUEST_TIMEOUT_SEC = 15;

export class SessionManager {
  private _sessionId: string | null = null;
  private _sessionTitle = 'JiuwenSwarm';
  private pending = new Map<string, PendingRequest>();
  private channelId: string;
  private sessionListeners: ((sid: string | null) => void)[] = [];

  constructor(private readonly ws: WsClient, channelId = 'ide') {
    this.channelId = channelId;
    this.ws.on('message', (msg) => this.handleMessage(msg));
  }

  get sessionId(): string | null { return this._sessionId; }
  get sessionTitle(): string { return this._sessionTitle; }

  addSessionListener(l: (sid: string | null) => void): void {
    this.sessionListeners.push(l);
  }

  removeSessionListener(l: (sid: string | null) => void): void {
    const idx = this.sessionListeners.indexOf(l);
    if (idx >= 0) this.sessionListeners.splice(idx, 1);
  }

  /** Returns available models and the currently active model name. */
  async listModels(): Promise<{ models: Record<string, unknown>[]; activeModel: string | null }> {
    const payload = await this.request('models.list', {});
    const models = (payload.models as Record<string, unknown>[]) || [];
    const activeModel = payload.active_model as string | undefined || null;
    return { models, activeModel };
  }

  /** List recent sessions */
  async listSessions(limit = 20): Promise<SessionInfo[]> {
    const payload = await this.request('session.list', { limit });
    const arr = payload.sessions as SessionInfo[] | undefined;
    if (!arr) return [];
    return arr.map((obj) => ({
      session_id: (obj.session_id as string) || '',
      title: obj.title as string | undefined,
      last_message_at: obj.last_message_at as number | undefined,
      created_at: obj.created_at as number | undefined,
      message_count: obj.message_count as number | undefined,
    }));
  }

  /** Switch to an existing session */
  async switchSession(sid: string): Promise<void> {
    const payload = await this.request('session.switch', { session_id: sid }, sid);
    this.setSessionId(sid);
    this._sessionTitle = (payload.title as string) || sid;
  }

  /**
   * Fire-and-forget chat message. Streaming events arrive via WsClient 'message' events.
   * If [ideContext] is non-null it is prepended to [content].
   * If [mediaItems] is non-null, image attachments are included in base64 format.
   */
  sendChat(
    content: string,
    mode: string,
    requestId: string,
    ideContext?: string,
    mediaItems?: unknown[],
  ): boolean {
    const sid = this._sessionId;
    if (!sid) return false;
    const fullContent = ideContext && ideContext.trim()
      ? `${ideContext}\n\n${content}`
      : content;
    const params: Record<string, unknown> = {
      content: fullContent,
      mode,
      session_id: sid,
    };
    if (mediaItems && mediaItems.length > 0) {
      params.media_items = mediaItems;
    }
    const msg = makeRequest('chat.send', params, this.channelId);
    msg.id = requestId;
    return this.ws.send(msg);
  }

  // ──────────────────────────────────────────
  // Internal
  // ──────────────────────────────────────────

  private request(method: string, params: Record<string, unknown>, sid?: string): Promise<Record<string, unknown>> {
    const p = { ...params };
    if (sid) p.session_id = sid;
    const msg = makeRequest(method, p, this.channelId);
    return new Promise((resolve, reject) => {
      const id = msg.id!;
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`Request '${method}' timed out`));
      }, REQUEST_TIMEOUT_SEC * 1000);

      this.pending.set(id, { resolve, reject, timer });
      if (!this.ws.send(msg)) {
        clearTimeout(timer);
        this.pending.delete(id);
        reject(new Error('WebSocket not connected'));
      }
    });
  }

  private handleMessage(msg: JiuwenMessage): void {
    // ── E2A format responses ──
    const responseKind = msg.payload?.response_kind as string | undefined;
    if (responseKind) {
      const rid = msg.payload?.request_id as string | undefined;
      if (!rid) return;
      const future = this.pending.get(rid);
      if (!future) return;
      if (responseKind === 'e2a.complete') {
        const body = (msg.payload?.body as Record<string, unknown>) || {};
        const result = (body.result as Record<string, unknown>) || body;
        clearTimeout(future.timer);
        this.pending.delete(rid);
        future.resolve(result);
      } else if (responseKind === 'e2a.error') {
        const body = (msg.payload?.body as Record<string, unknown>) || {};
        const err = (body.message as string) || 'E2A error';
        clearTimeout(future.timer);
        this.pending.delete(rid);
        future.reject(new Error(err));
      } else {
        clearTimeout(future.timer);
        this.pending.delete(rid);
        future.reject(new Error(`Unexpected response_kind: ${responseKind}`));
      }
      return;
    }

    // ── Gateway legacy format responses ──
    if (msg.type === 'res' && msg.id) {
      const future = this.pending.get(msg.id);
      if (future) {
        clearTimeout(future.timer);
        this.pending.delete(msg.id);
        if (msg.ok) {
          future.resolve((msg.payload as Record<string, unknown>) || {});
        } else {
          const err =
            (msg.payload as Record<string, unknown> | undefined)?.error as string | undefined
            || 'Request failed';
          future.reject(new Error(err));
        }
      }
      return;
    }

    // ── Events (connection.ack gives us the session) ──
    if (msg.type === 'event' && msg.event) {
      const payload = (msg.payload as Record<string, unknown>) || {};
      if (msg.event === 'connection.ack') {
        const sid = payload.session_id as string | undefined;
        if (sid) {
          this.setSessionId(sid);
        }
      }
    }
  }

  private setSessionId(sid: string | null): void {
    if (this._sessionId === sid) return;
    this._sessionId = sid;
    this.sessionListeners.forEach((l) => l(sid));
  }
}

File: packages/jetbrains-plugin/src/main/kotlin/com/jiuwenswarm/plugin/client/SessionManager.kt
package com.jiuwenswarm.plugin.client

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.intellij.openapi.diagnostic.logger
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

private val LOG = logger<SessionManager>()
private val gson = Gson()
private const val REQUEST_TIMEOUT_SEC = 15L

data class SessionInfo(
    val session_id: String,
    val title: String?,
    val last_message_at: Long?,
    val created_at: Long?,
    val message_count: Int?,
)

class SessionManager(
    private val ws: WsClient,
    private val channelId: String = "ide",
) {
    var sessionId: String? = null
        private set
    var sessionTitle: String = "JiuwenSwarm"
        private set

    private val pending = ConcurrentHashMap<String, CompletableFuture<JsonObject>>()
    private val listener: MessageListener = { msg -> handleMessage(msg) }

    init {
        ws.addMessageListener(listener)
    }

    fun dispose() {
        ws.removeMessageListener(listener)
        pending.values.forEach { it.cancel(true) }
        pending.clear()
    }

    /** Create a new session, blocking up to [REQUEST_TIMEOUT_SEC] seconds */
    fun createSession(): String {
        val payload = request("session.create", emptyMap())
        // Server uses camelCase "sessionId" in the response payload
        val sid = payload.get("sessionId")?.asString
            ?: payload.get("session_id")?.asString
            ?: error("No sessionId in response")
        sessionId = sid
        sessionTitle = payload.get("title")?.asString ?: "New Session"
        return sid
    }

    fun listSessions(limit: Int = 20): List<SessionInfo> {
        val payload = request("session.list", mapOf("limit" to limit))
        val arr = payload.getAsJsonArray("sessions") ?: return emptyList()
        return arr.map { el ->
            val obj = el.asJsonObject
            SessionInfo(
                session_id = obj.get("session_id")?.asString ?: "",
                title = obj.get("title")?.asString,
                last_message_at = obj.get("last_message_at")?.asLong,
                created_at = obj.get("created_at")?.asLong,
                message_count = obj.get("message_count")?.asInt,
            )
        }
    }

    fun switchSession(sid: String) {
        val payload = request("session.switch", mapOf("session_id" to sid), sid)
        sessionId = sid
        sessionTitle = payload.get("title")?.asString ?: sid
    }

    /** Fire-and-forget chat message. Streaming events arrive via WsClient message listener. */
    fun sendChat(content: String, mode: String, requestId: String): Boolean {
        val sid = sessionId ?: return false
        val msg = buildJsonObject {
            addProperty("id", requestId)
            addProperty("type", "req")
            addProperty("channel_id", channelId)
            addProperty("session_id", sid)
            addProperty("req_method", "chat.send")
            add("params", buildJsonObject {
                addProperty("content", content)
                addProperty("mode", mode)
            })
            addProperty("timestamp", System.currentTimeMillis() / 1000.0)
        }
        return ws.send(msg)
    }

    // ──────────────────────────────────────────
    private fun request(
        method: String,
        params: Map<String, Any>,
        sid: String? = null,
    ): JsonObject {
        val id = UUID.randomUUID().toString()
        val msg = buildJsonObject {
            addProperty("request_id", id)
            addProperty("type", "req")
            addProperty("channel_id", channelId)
            if (sid != null) addProperty("session_id", sid)
            addProperty("req_method", method)
            add("params", gson.toJsonTree(params).asJsonObject)
            addProperty("timestamp", System.currentTimeMillis() / 1000.0)
        }
        val future = CompletableFuture<JsonObject>()
        pending[id] = future
        if (!ws.send(msg)) {
            pending.remove(id)
            throw IllegalStateException("WebSocket not connected")
        }
        return try {
            future.get(REQUEST_TIMEOUT_SEC, TimeUnit.SECONDS)
        } catch (e: Exception) {
            pending.remove(id)
            throw RuntimeException("Request '$method' failed: ${e.message}", e)
        }
    }

    private fun handleMessage(msg: JsonObject) {
        // ── E2A format responses (server always sends these) ──
        val requestId = msg.get("request_id")?.asString
        val status = msg.get("status")?.asString
        if (requestId != null && status != null) {
            val future = pending.remove(requestId) ?: return
            if (status == "succeeded") {
                val body = msg.getAsJsonObject("body")
                // The actual result is nested under body.result
                val result = body?.getAsJsonObject("result") ?: JsonObject()
                future.complete(result)
            } else {
                val body = msg.getAsJsonObject("body")
                val error = body?.get("message")?.asString ?: "Request failed"
                future.completeExceptionally(RuntimeException(error))
            }
            return
        }

        // ── Legacy format responses (fallback, unlikely with current server) ──
        if (msg.get("type")?.asString == "res") {
            val rid = msg.get("request_id")?.asString ?: return
            val future = pending.remove(rid) ?: return
            if (msg.get("ok")?.asBoolean == true) {
                val payload = msg.getAsJsonObject("payload") ?: JsonObject()
                future.complete(payload)
            } else {
                val err = msg.getAsJsonObject("payload")?.get("error")?.asString ?: "Request failed"
                future.completeExceptionally(RuntimeException(err))
            }
        }
    }
}

// ──────────────────────────────────────────────────────────────────
// Tiny DSL to build JsonObject without Kotlin serialization overhead
// ──────────────────────────────────────────────────────────────────
private fun buildJsonObject(block: JsonObject.() -> Unit): JsonObject =
    JsonObject().also(block)
